tests/internal/test_module.py

```diff
@@ -0,0 +1,269 @@
+from os.path import dirname
+import sys
+
+import mock
+import pytest
+
+from ddtrace.internal.module import ModuleWatchdog
+from ddtrace.internal.module import origin
+import tests.test_module
+
+
+@pytest.fixture
+def module_watchdog():
+    ModuleWatchdog.install()
+
+    assert ModuleWatchdog.is_installed()
+
+    yield ModuleWatchdog
+
+    ModuleWatchdog.uninstall()
+
+
+def test_watchdog_install_uninstall():
+    assert not isinstance(sys.modules, ModuleWatchdog)
+    ModuleWatchdog.install()
+    assert isinstance(sys.modules, ModuleWatchdog)
+    ModuleWatchdog.uninstall()
+    assert not isinstance(sys.modules, ModuleWatchdog)
+
+
+def test_import_origin_hook_for_imported_module(module_watchdog):
+    hook = mock.Mock()
+    module = sys.modules[__name__]
+    module_watchdog.register_origin_hook(origin(module), hook)
+
+    hook.assert_called_once_with(module)
+
+
+def test_import_module_hook_for_imported_module(module_watchdog):
+    hook = mock.Mock()
+    module = sys.modules[__name__]
+    module_watchdog.register_module_hook(module.__name__, hook)
+
+    hook.assert_called_once_with(module)
+
+
+def test_register_hook_without_install():
+    with pytest.raises(RuntimeError):
+        ModuleWatchdog.register_origin_hook(__file__, mock.Mock())
+
+    with pytest.raises(RuntimeError):
+        ModuleWatchdog.register_module_hook(__name__, mock.Mock())
+
+
+@pytest.mark.subprocess(env=dict(MODULE_ORIGIN=origin(tests.test_module)))
+def test_import_origin_hook_for_module_not_yet_imported():
+    import os
+    import sys
+
+    from mock import mock
+
+    from ddtrace.internal.module import ModuleWatchdog
+
+    name = "tests.test_module"
+    path = os.getenv("MODULE_ORIGIN")
+    hook = mock.Mock()
+
+    ModuleWatchdog.install()
+
+    ModuleWatchdog.register_origin_hook(path, hook)
+
+    hook.assert_not_called()
+    assert path in ModuleWatchdog._instance._hook_map
+    assert name not in sys.modules
+
+    # Check that we are not triggering hooks on the wrong module
+    import tests.internal  # noqa
+
+    hook.assert_not_called()
+
+    # We import multiple times to check that the hook is called once only
+    __import__(name)
+    __import__(name)
+
+    assert name in sys.modules
+
+    hook.assert_called_once_with(sys.modules[name])
+
+    ModuleWatchdog.uninstall()
+
+
+@pytest.mark.subprocess
+def test_import_module_hook_for_module_not_yet_imported():
+    import sys
+
+    from mock import mock
+
+    from ddtrace.internal.module import ModuleWatchdog
+
+    name = "tests.test_module"
+    hook = mock.Mock()
+
+    ModuleWatchdog.install()
+
+    ModuleWatchdog.register_module_hook(name, hook)
+
+    hook.assert_not_called()
+    assert name not in sys.modules
```

Comment:
Can a similar check be added here as was added above for after an import?

Reply:
Ah, of course! 🙇